### PR TITLE
[RPS-20] Add attachment to a publication

### PR DESCRIPTION
**Pull request by ziemsky**

### CODE REVIEW

File: repository-data/webfiles/src/main/resources/site/freemarker/publication/publication.ftl [hunk 1]
```diff
@@ -7,6 +8,19 @@
...
+
+    <div>
+        <h4>Attachments:</h4>
+        <ul>
+        <#if document.attachments??>
+                <#list document.attachments as attachment>
+                    <li><a href="<@hst.link hippobean=attachment/>">${attachment.filename}</a></li>
+                </#list>
+        <#else>
+            <li>No attachments found.</li>
```
Comment (ivorc) [inline repository-data/webfiles/src/main/resources/site/freemarker/publication/publication.ftl]: Suggest just "No attachments"

File: site/src/main/webapp/WEB-INF/web.xml [hunk 2]
```diff
@@ -108,6 +108,18 @@
   <servlet>
     <servlet-name>BinariesServlet</servlet-name>
     <servlet-class>org.hippoecm.hst.servlet.BinariesServlet</servlet-class>
+
+    <init-param>
+        <param-name>contentDispositionContentTypes</param-name>
+        <param-value>*/*</param-value>
```
Comment (ivorc) [inline site/src/main/webapp/WEB-INF/web.xml]: * Should PDFs open in new tab?
* What is max file size we should allow? Is this is in scope?
* What file types are allowed? Is this is in scope?
* Do we need to show file size / type?

Comment (ziemsky) [reply site/src/main/webapp/WEB-INF/web.xml]: Questions forwarded to the product owner.

File: acceptance-tests/src/test/java/uk/nhs/digital/ps/test/acceptance/util/FileHelper.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import static java.lang.System.lineSeparator;
+import static java.nio.file.Files.readAllLines;
+import static java.nio.file.Files.write;
+import static java.util.stream.Collectors.joining;
+
+public class FileHelper {
+
+    public static Path createTempFile(final String content) {
+        try {
+            final Path tempFile = Files.createTempFile("ps-hippo-acc-test-", ".tmp");
```
[the remark this answers is not in the document]
Comment (ziemsky) [reply acceptance-tests/src/test/java/uk/nhs/digital/ps/test/acceptance/util/FileHelper.java]: Done.

File: acceptance-tests/src/test/java/uk/nhs/digital/ps/test/acceptance/util/FileHelper.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+            tempFile.toFile().deleteOnExit();
+
+            final String randomContent = content;
```
[the remark this answers is not in the document]
Comment (ziemsky) [reply acceptance-tests/src/test/java/uk/nhs/digital/ps/test/acceptance/util/FileHelper.java]: Gone as a result with a wider refactoring - content is now provided from outside.

File: acceptance-tests/src/test/java/uk/nhs/digital/ps/test/acceptance/models/Attachment.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
+package uk.nhs.digital.ps.test.acceptance.models;
+
+public class Attachment {
+    private String name;
+    private String content;
+
+    public Attachment(final String name, final String content) {
```
[the remark this answers is not in the document]
Comment (ziemsky) [reply acceptance-tests/src/test/java/uk/nhs/digital/ps/test/acceptance/models/Attachment.java]: Done. The content used to populate the file is now a random array of bytes rather than string.

File: acceptance-tests/src/test/java/uk/nhs/digital/ps/test/acceptance/pages/ConsumablePublicationPage.java [hunk 6]
```diff
@@ -31,4 +31,15 @@ public String getSummary() {
...
 
+    public WebElement getAttachmentElement() {
+        return helper.findElement(By.tagName("a"));
+    }
+
+    public String getAttachmentName() {
+        return getAttachmentElement().getText();
+    }
+
+    public String getAttachmentUrl() {
```
[the remark this answers is not in the document]
Comment (ziemsky) [reply acceptance-tests/src/test/java/uk/nhs/digital/ps/test/acceptance/pages/ConsumablePublicationPage.java]: Now gone.

File: acceptance-tests/src/test/java/uk/nhs/digital/ps/test/acceptance/steps/ContentSteps.java [hunk 7]
```diff
@@ -3,16 +3,29 @@
...
 import uk.nhs.digital.ps.test.acceptance.pages.ConsumablePublicationPage;
 import uk.nhs.digital.ps.test.acceptance.pages.ContentPage;
 
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
 import static org.hamcrest.CoreMatchers.is;
 import static org.junit.Assert.assertThat;
+import static org.slf4j.LoggerFactory.getLogger;
+import static uk.nhs.digital.ps.test.acceptance.util.FileHelper.read;
```
[the remark this answers is not in the document]
Comment (ziemsky) [reply acceptance-tests/src/test/java/uk/nhs/digital/ps/test/acceptance/steps/ContentSteps.java]: The two statically imported methods have been renamed to a more descriptive `readFileAsByteArray` and `waitUntilFileAppears`.

As discussed, static imports were left used in the interest of brevity/readability where there was less risk of ambiguity and removed where qualifying them with the class' name made the code more clear.

File: repository-data/webfiles/src/main/resources/site/freemarker/publication/publication.ftl (hunk 1 above)
Comment (ziemsky) [reply repository-data/webfiles/src/main/resources/site/freemarker/publication/publication.ftl]: Corrected.

File: site/src/main/webapp/WEB-INF/web.xml (hunk 2 above)
Comment (ziemsky) [reply site/src/main/webapp/WEB-INF/web.xml]: And response now documented by BA in JIRA and implemented (allowed file types and max size configured, file sizes displayed; support for extra large files deferred to a separate, future story).

**Comment by ngta1** [comment]
Reviewed with Rafal. All looks good.
